oci/build-layer.py

```diff
@@ -0,0 +1,299 @@
+#!/usr/bin/env python3
+
+from datetime import datetime, timezone
+from pathlib import PurePosixPath, PosixPath
+import io
+import json
+import os
+import re
+import subprocess
+import sys
+import tarfile
+
+
+def parse_path(name):
+    path = PurePosixPath(name)
+    if path.is_absolute():
+        path = path.relative_to('/')
+
+    if path.as_posix() == '.':
+        return '.'
+    else:
+        return './{}'.format(path)
+
+
+number_regex = re.compile('[0-9]+')
+mode_regex = re.compile('[0-9]{4}')
+
+
+def parse_mode(mode):
+    if mode == None:
+        return ModeSet(None)
+    elif mode_regex.fullmatch(mode):
+        return ModeSet(int(mode, 8))
+    elif number_regex.fullmatch(mode):
+        raise Exception(
+            "Invalid mode: {}, please specify numeric modes as 4-digit octal".
+            format(mode))
+    else:
+        return ModeChange(mode)
+
+
+def assert_number(num):
+    if isinstance(num, int):
+        return num
+    else:
+        raise Exception("Got {}, but should be an int")
+
+
+class ModeSet:
+    def __init__(self, mode):
+        self.mode = mode
+
+    def apply(self, mode, is_dir=False):
+        if self.mode == None:
+            return mode
+        else:
+            return self.mode
+
+
+# Support symbolic mode changes
```

Comment:
I'll admit I spent way more time than I should have implementing this but I felt it was necessary for some of our use cases copying in a derivation and making it writable for example.